controllers/billingentity_email_cronjob.go

```diff
@@ -0,0 +1,112 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/multierr"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/appuio/control-api/mailsenders"
+	"github.com/prometheus/client_golang/prometheus"
+
+	billingv1 "github.com/appuio/control-api/apis/billing/v1"
+)
+
+// BillingEntityEmailCronJob periodically checks billing entities and sends notification emails if appropriate
+type BillingEntityEmailCronJob struct {
+	client.Client
+
+	Recorder record.EventRecorder
+	Scheme   *runtime.Scheme
+
+	MailSender mailsenders.MailSender
+
+	mailRecipientAddress string
+
+	failureCounter prometheus.Counter
+	successCounter prometheus.Counter
+}
+
+func NewBillingEntityEmailCronJob(client client.Client, eventRecorder record.EventRecorder, scheme *runtime.Scheme, mailSender mailsenders.MailSender, MailRecipientAddress string) BillingEntityEmailCronJob {
+	return BillingEntityEmailCronJob{
+		Client:               client,
+		Recorder:             eventRecorder,
+		MailSender:           mailSender,
+		mailRecipientAddress: MailRecipientAddress,
+		failureCounter:       newFailureCounter("control_api_billingentity_emails"),
+		successCounter:       newSuccessCounter("control_api_billingentity_emails"),
+	}
+
+}
+
+func (r *BillingEntityEmailCronJob) GetMetrics() prometheus.Collector {
+	reg := prometheus.NewRegistry()
+	reg.MustRegister(r.failureCounter)
+	reg.MustRegister(r.successCounter)
+	return reg
+}
+
+//+kubebuilder:rbac:groups="rbac.appuio.io",resources=billingentities,verbs=get;list
+//+kubebuilder:rbac:groups="user.appuio.io",resources=billingentities,verbs=get;list
+//+kubebuilder:rbac:groups="rbac.appuio.io",resources=billingentities/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups="user.appuio.io",resources=billingentities/status,verbs=get;update;patch
+
+// Run lists all BillingEntity resources and sends notification emails if needed.
+func (r *BillingEntityEmailCronJob) Run(ctx context.Context) error {
+	log := log.FromContext(ctx)
+	log.V(1).Info("Sending Billing Entity E-Mails...")
+
+	list := &billingv1.BillingEntityList{}
+
+	err := r.Client.List(ctx, list)
+
+	if err != nil {
+		return err
+	}
+
+	var errors []error
+	for _, be := range list.Items {
```

Comment:
I wonder if it would make more sense to collect all billing entities for which we need to send an email, and send a single email which contains all the billing entities that have been created or updated since the last email was sent.

That way we avoid spamming the recipient regardless of the amount of changes that happened.

Obviously, there's a tradeoff between amount of emails being sent and the blast radius of a single sending failure. A compromise might be to retry sending on transient errors (if we can identify those) if we have a single email per cron execution.

Reply:
Ask the recipients what their preference is.

Reply:
As per the discussion in the daily, we'll move forward with this, and we can always change it later if need be.

Reply:
I got confirmation that a single email per change is preferred.